ably/proto_presence_message.go

```diff
@@ -58,3 +60,45 @@ func (m PresenceMessage) String() string {
 		"update",
 	}[m.Action], m.ClientID, m.Data)
 }
+
+func (msg *PresenceMessage) isServerSynthesized() bool {
+	return !strings.HasPrefix(msg.ID, msg.ConnectionID)
+}
+
+func (msg *PresenceMessage) getMsgSerialAndIndex() (int64, int64, error) {
+	msgIds := strings.Split(msg.ID, ":")
+	if len(msgIds) != 3 {
+		return 0, 0, fmt.Errorf("parsing error, the presence message has invalid id %v", msg.ID)
+	}
+	msgSerial, err := strconv.ParseInt(msgIds[1], 10, 64)
+	if err != nil {
+		return 0, 0, fmt.Errorf("parsing error, the presence message has invalid msgSerial, for msgId %v", msg.ID)
+	}
+	msgIndex, err := strconv.ParseInt(msgIds[2], 10, 64)
+	if err != nil {
+		return 0, 0, fmt.Errorf("parsing error, the presence message has invalid msgIndex, for msgId %v", msg.ID)
+	}
+	return msgSerial, msgIndex, nil
+}
+
+// RTP2b, RTP2c
+func (msg1 *PresenceMessage) IsNewerThan(msg2 *PresenceMessage) (bool, error) {
```

Comment:
Need to check spec what happens when values are equal. There is an explicit mention to choose one over another.

Reply:
updated as per spec